fetch/build formula from source when modified

dunn opened the pull request:
Closes #36068.

Is performance something to worry about here?

dunn commented:
Oh, forgot to add the override flag @DylanLacey mentioned.

dunn commented:
Nevermind, `--force-bottle` still works to override it.

dunn commented:
```
  1) Error:
BottleHookTests#test_has_no_bottle:
NoMethodError: undefined method `is_modified?' for #<BottleHookTests::FormulaDouble:0x007f9203054078>
    /usr/local/Library/Homebrew/formula_installer.rb:65:in `pour_bottle?'
```
This is weird since formula_installer `requires` formula.  I must be missing something.

DylanLacey commented:
Oh bugger.  This'll teach me to write an implementation on a RailsCamp and forget to send a PR >.>

dunn commented:
I mean if yours is better I'd be happy to close this in favor of it.

xu-cheng commented on Library/Homebrew/formula.rb:
```diff
@@ -595,6 +595,14 @@ def inspect
+    old_wd = Pathname.pwd
+    Dir.chdir path.dirname
+    diff = Utils.popen_read("git", "diff", "--diff-filter=M", "--name-only", "#{path}")
+    Dir.chdir old_wd
```
You can do:
```ruby
path.parent.cd do
  diff = ...
end
```

Another thing that you need to check is formula file may be outside git repo.

dunn replied:
> Another thing that you need to check is formula file may be outside git repo.

That was why I added the `exitstatus` check; git should return non-zero in those cases. Are there cases where that's insufficient?

DylanLacey commented:
@Dunn I don't believe so;  I pushed it as https://github.com/Homebrew/homebrew/pull/41433/files.  I just was chastising myself for having and implementing a reasonable idea, then not actually going 100% through forgetfulness and lack of tests :P

tdsmith commented:
I do think it's important that a solution announces "Formula modified locally; building from source" or similar.

dunn commented:
Updated; but let me know if this should be closed in favor of @DylanLacey's.

xu-cheng commented on Library/Homebrew/formula.rb:
```diff
@@ -595,6 +595,13 @@ def inspect
+  def modified?
+    path.parent.cd do
+      diff = Utils.popen_read("git", "diff", "--diff-filter=M", "--name-only", "#{path}")
+      return !diff.empty? && $?.exitstatus == 0
```
`return` is redundant

xu-cheng commented on Library/Homebrew/formula.rb:
```diff
@@ -595,6 +595,13 @@ def inspect
     s << ") #{path}>"
   end
 
+  def modified?
```
Let's add `return false unless which("git")` to check whether git is available.

mikemcquaid replied:
Let's also call this `file_modified?`

mikemcquaid commented on Library/Homebrew/formula_installer.rb:
```diff
@@ -62,6 +62,11 @@ def pour_bottle? install_bottle_options={:warn=>false}
     return true  if force_bottle? && bottle
+    if formula.modified?
+      ohai "#{formula.full_name} is modified; building from source"
+      ohai "Build from bottle with --force-bottle"
```
`Install from a bottle`

mikemcquaid commented on Library/Homebrew/cmd/fetch.rb:
```diff
@@ -32,6 +32,11 @@ def fetch
     return false unless f.bottle && f.pour_bottle?
+    if formula.modified?
+      ohai "#{formula.full_name} is modified; fetching source"
+      ohai "Fetch bottle with --force-bottle"
```
`Fetch the bottle with`

dunn commented:
Still need help interpreting the test failure, I'm afraid.

mikemcquaid commented on Library/Homebrew/formula.rb:
```diff
@@ -595,6 +595,14 @@ def inspect
+      diff = Utils.popen_read("git", "diff", "--diff-filter=M", "--name-only", "#{path}")
+      !diff.empty? && $?.exitstatus == 0
+    end
+  end
```
Also add this method here: https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/test/test_bottle_hooks.rb#L6-L10

xu-cheng commented on Library/Homebrew/cmd/fetch.rb:
```diff
@@ -32,6 +32,11 @@ def fetch
     return false unless f.bottle && f.pour_bottle?
+    if formula.file_modified?
+      ohai "#{formula.full_name} is modified; fetching source"
+      ohai "Fetch the bottle with --force-bottle"
```
This is kinda misleading. It would be better to be `To fetch the bottle, run with --force-bottle`. Also use `puts` for the second `ohai`.

dunn replied:
Using a `puts` means we're not restricted by length, so maybe even "To fetch the bottle instead, run with --force-bottle"?

mikemcquaid replied:
That sounds good.

xu-cheng commented on Library/Homebrew/formula_installer.rb:
```diff
@@ -62,6 +62,11 @@ def pour_bottle? install_bottle_options={:warn=>false}
     return true  if force_bottle? && bottle
+    if formula.file_modified?
+      ohai "#{formula.full_name} is modified; building from source"
+      ohai "Install from a bottle with --force-bottle"
```
Similar problem as above.

dunn commented:
OK, I think that covers everything.

dunn commented:
Not getting those test failures locally.

mikemcquaid commented:
I currently see this message printed twice (it should only be output called once). Once that's fixed we're good to go.

dunn commented:
Sorry, where are you seeing doubled output?

Also, added `.rb` to the `ohai`, since it's the file that's been modified.

mikemcquaid commented:
If I run `brew install wget` after modifying the formula.

dunn commented:
Oh, I was testing with `brew install when` and it doesn't happen for that formula. Bizarre; I'll investigate.

dunn commented:
It makes sense that we get the message twice for formula like libtorrent which call `pour_bottle?`, but I'm not sure why wget does it.

dunn commented:
Anyway, fixed!

mikemcquaid commented on Library/Homebrew/cmd/fetch.rb:
```diff
@@ -32,6 +32,11 @@ def fetch
     return true if ARGV.force_bottle? && f.bottle
     return false unless f.bottle && f.pour_bottle?
+    if formula.file_modified?
+      ohai "#{formula.full_name}.rb is modified; fetching source"
```
`formula.path.to_s.gsub("#{HOMEBREW_PREFIX}/", '')` might be a bit better (particularly for stuff from taps).

dunn replied:
On narrow tmux/screen panes the path will get cut off, but I think that's fine:
![screenshot 2015-07-11 10 25 18](https://cloud.githubusercontent.com/assets/985416/8634688/b62f0be4-27b7-11e5-9f98-d5b3a658a782.png)

mikemcquaid replied:
Maybe go for `PATH is modified!` and then putting the rest fitting on the next line? What do you think?

mikemcquaid commented on Library/Homebrew/formula_installer.rb:
```diff
@@ -180,6 +180,10 @@ def install
     unless @poured_bottle
+      if formula.file_modified?
+        ohai "#{formula.full_name}.rb is modified; building from source"
+        puts "To install from a bottle instead, run with --force-bottle"
```
I wonder if there's somewhere you can unify the printing of this message in a method somewhere (just so we don't need to manually keep the two strings in sync).

dunn replied:
They're not quite the same, though. Unless you'd like them to be?

mikemcquaid commented:
@dunn This is looking awesome. A few more tweaks; almost there!

dunn commented:
Three changes:
- Fixed the `git diff` to detect committed as well as staged changes
- changed the `ohai` to `opoo` (I think it should stand out a bit from all the other `ohai`s, but either way is OK, really)
- Modified the message as below:
![screenshot 2015-07-11 14 03 21](https://cloud.githubusercontent.com/assets/985416/8635402/ef520a84-27d5-11e5-9b90-2a9906ab6aca.png)
